### ISSUE TITLE
Strategies for COW  friendly code page sharing

### ISSUE THREAD

**Issue by andrewchambers** [janet-lang/janet#475]
Janet web servers, especially with large amounts of bytecode and embedded resources would benefit greatly if there were ways to share this memory across VM process instances.

A typical way to do this, is to start your application, listen on a socket, then fork N times with each fork is able to accept and handle connections. This way the loaded embedded resources and bytecode are all stored in shared COW pages in memory.

Unfortunately the current janet VM trashes these COW pages by writing the mark bit during GC. 
Ruby (mostly?) solved this in 2.0 by moving the mark bits into a compressed bitmap in this patch:

http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-core/41916

This is one way to solve it, though perhaps not the only way.

Another way may be a special gc bit marking all live immutable and non reference containing gc objects  as permanently marked - they will never be collected and never need to be walked again.

```
(defn main [...]
  (init-application)
  (gccollect)
  (mark-objects-with-magic-bit)
  (fork-and-serve))
```

This solution also makes GC more efficient for an upfront cost, as we can skip all these objects in future collections. I'm unsure if there is an efficient way to compute the relevant object subgraphs.

**Comment by andrewchambers** [comment]
Just a note, on linux its possible to use a program called 'smem' to view page sharing in experiments. I also did some experiments using github.com/andrewchambers/janet-fork , which may be useful for testing.

**Comment by bakpakin** [comment]
This is definitely something I will think about, but exactly what solution we use  is definitely up for debate. The magic bit solution definitely seems the simplest to add.

**Comment by andrewchambers** [comment]
another thing to consider for this is if it allows the os to page out infrequently used bytecode.

**Comment by cody271** [comment]
> Janet web servers, especially with large amounts of bytecode and embedded resources would benefit greatly (in the form of reduced memory overhead) if there were ways to share this memory across VM process instances.
>
This is a builtin feature of LuaJIT, and a huge factor in its really high performance in some larger codebases.

Basic example with a Lua module named `foo`:
```bash
$ cat foo.lua 
```
```lua
return { bar = function() return 42 end }
```
```bash
$ luajit -e 'print(require("foo").bar())' 
42
```
Use LuaJIT to compile `foo.lua` into a native module `foo_cow.so`:
```bash
$ luajit -b -t c -n foo_cow foo.lua foo_cow.c
$ gcc -fPIC -shared foo_cow.c -o foo_cow.so
$ nm foo_cow.so | grep lua
luaJIT_BC_foo_cow
```
Which can now be shared across VM process instances:
```bash
$ luajit -e 'print(require("foo_cow").bar())'
42
```
This is simpler (and safer) than having to modify the GC or memory paging behavior.
Unless I'm misunderstanding, this same technique should work for Janet as well.

**Comment by andrewchambers** [comment]
> This is simpler (and safer) than having to modify the GC or memory paging behavior.

I am not sure it is simpler as it would require writing a janet bytecode to native code compiler, it is an alternative and cool approach.

**Comment by cody271** [comment]
> I am not sure it is simpler as it would require writing a janet bytecode to native code compiler, it is still an alternative and cool approach.

No no much simpler, no native AOT. Just embeds the bytecode as an exported symbol in the native lib:

```bash
$ cat foo_cow.c
```
```c
#ifdef __cplusplus
extern "C"
#endif
#ifdef _WIN32
__declspec(dllexport)
#endif
const unsigned char luaJIT_BC_foo_cow[] = {
27,76,74,2,10,15,0,0,1,0,0,0,2,41,0,42,0,76,0,2,0,35,3,0,2,0,3,0,5,53,0,1,0,
51,1,0,0,61,1,2,0,50,0,0,128,76,0,2,0,8,98,97,114,1,0,0,0,0
};
```

**Comment by andrewchambers** [comment]
This still needs a magic bit in the gc since the embedded bytecode would need to be marked as special. This would be a good feature to add to jpm generated binaries.

**Comment by cody271** [comment]
> This still needs modifications in the gc since the embedded bytecode would need to be marked as special so it isn't freed.
>
Yes this is an internal implementation detail though. _Very_ different than exposing a new user facing API that can manipulate the GC. I would caution against exposing something like `(mark-objects-with-magic-bit)`.

**Comment by andrewchambers** [comment]
Oh yeah definitely, the way ruby handles it is an internal detail too though.
